File: blueprints/pwa.py
"""
Blueprint PWA - Mi Agente Viajes

Rutas:
- /offline - Página offline para PWA
- /sw.js - Service Worker (servido desde static pero con scope /)
"""

from flask import Blueprint, render_template, send_from_directory, current_app
import os

pwa_bp = Blueprint('pwa', __name__)


@pwa_bp.route('/offline')
def offline():
    """Página mostrada cuando no hay conexión."""
    return render_template('offline.html')


@pwa_bp.route('/sw.js')
def service_worker():
    """
    Servir Service Worker desde raíz para scope completo.

    El SW debe servirse desde / para tener scope sobre toda la app.
    Si se sirve desde /static/, el scope estaría limitado a /static/.
    """
    return send_from_directory(
        os.path.join(current_app.root_path, 'static'),
        'sw.js',
        mimetype='application/javascript'
    )


@pwa_bp.route('/apple-touch-icon.png')
def apple_touch_icon():
    """Servir apple-touch-icon desde raíz para compatibilidad iOS."""
    return send_from_directory(
        os.path.join(current_app.root_path, 'static', 'icons'),
        'apple-touch-icon.png',
        mimetype='image/png'
    )


@pwa_bp.route('/firebase-messaging-sw.js')
def firebase_messaging_sw():
    """Servir Firebase Messaging Service Worker desde raíz."""
    return send_from_directory(
        os.path.join(current_app.root_path, 'static'),
        'firebase-messaging-sw.js',
        mimetype='application/javascript'
    )


@pwa_bp.route('/force-update')
def force_update():
    """
    Página especial que desregistra el Service Worker y limpia todo el cache.
    Usar cuando el SW está corrupto o no se actualiza.
    """
    return """
<!DOCTYPE html>
<html lang="es">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Forzar Actualización - Mi Agente Viajes</title>
    <style>
        * { margin: 0; padding: 0; box-sizing: border-box; }
        body {
            font-family: -apple-system, BlinkMacSystemFont, 'SF Pro Display', sans-serif;
            background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
            display: flex;
            align-items: center;
            justify-content: center;
            min-height: 100vh;
            padding: 24px;
        }
        .container {
            background: white;
            border-radius: 20px;
            padding: 40px;
            max-width: 500px;
            text-align: center;
            box-shadow: 0 20px 60px rgba(0,0,0,0.3);
        }
        h1 {
            font-size: 2rem;
            margin-bottom: 16px;
            color: #1D1D1F;
        }
        .status {
            font-size: 1.125rem;
            color: #666;
            margin-bottom: 24px;
            min-height: 60px;
        }
        .spinner {
            width: 50px;
            height: 50px;
            border: 4px solid #E5E5E7;
            border-top-color: #667eea;
            border-radius: 50%;
            animation: spin 0.8s linear infinite;
            margin: 0 auto 24px;
        }
        @keyframes spin {
            to { transform: rotate(360deg); }
        }
        .success {
            color: #34C759;
            font-size: 4rem;
            margin-bottom: 16px;
        }
        button {
            background: #667eea;
            color: white;
            border: none;
            padding: 14px 32px;
            border-radius: 12px;
            font-size: 1.125rem;
            font-weight: 600;
            cursor: pointer;
            transition: all 0.2s;
        }
        button:hover {
            background: #5568d3;
            transform: translateY(-2px);
            box-shadow: 0 4px 12px rgba(102, 126, 234, 0.4);
        }
        button:disabled {
            background: #ccc;
            cursor: not-allowed;
            transform: none;
        }
        .log {
            background: #F5F5F7;
            border-radius: 12px;
            padding: 16px;
            margin-top: 24px;
            font-family: monospace;
            font-size: 0.9rem;
            text-align: left;
            max-height: 200px;
            overflow-y: auto;
            color: #1D1D1F;
        }
        .log div {
            margin-bottom: 4px;
        }
    </style>
</head>
<body>
    <div class="container">
        <h1>🔄 Actualización Forzada</h1>
        <div class="spinner" id="spinner"></div>
        <div class="status" id="status">Limpiando cache...</div>
        <div class="log" id="log"></div>
        <button id="homeBtn" style="display: none;" onclick="window.location.href='/'">
            Volver a Inicio
        </button>
    </div>

    <script>
        const log = (msg) => {
            console.log(msg);
            const logEl = document.getElementById('log');
            const entry = document.createElement('div');
            entry.textContent = `[${new Date().toLocaleTimeString()}] ${msg}`;
            logEl.appendChild(entry);
            logEl.scrollTop = logEl.scrollHeight;
        };

        const updateStatus = (msg) => {
            document.getElementById('status').textContent = msg;
        };

        const showSuccess = () => {
            document.getElementById('spinner').style.display = 'none';
            document.getElementById('status').innerHTML = '<div class="success">✓</div>Actualización completa';
            document.getElementById('homeBtn').style.display = 'inline-block';
        };

        async function forceUpdate() {
            try {
                log('Iniciando actualización forzada...');

                // 1. Desregistrar Service Workers
                if ('serviceWorker' in navigator) {
                    updateStatus('Desregistrando Service Workers...');
                    const registrations = await navigator.serviceWorker.getRegistrations();
                    log(`Encontrados ${registrations.length} Service Workers`);

                    for (const registration of registrations) {
                        await registration.unregister();
                        log(`✓ SW desregistrado: ${registration.scope}`);
                    }
                }

                // 2. Limpiar todos los caches
                updateStatus('Limpiando caches...');
                if ('caches' in window) {
                    const cacheNames = await caches.keys();
                    log(`Encontrados ${cacheNames.length} caches`);

                    for (const name of cacheNames) {
                        await caches.delete(name);
                        log(`✓ Cache eliminado: ${name}`);
                    }
                }

                // 3. Limpiar localStorage y sessionStorage
                updateStatus('Limpiando storage...');
                try {
                    localStorage.clear();
                    sessionStorage.clear();
                    log('✓ LocalStorage y SessionStorage limpiados');
                } catch (e) {
                    log('⚠️ No se pudo limpiar storage: ' + e.message);
                }

                // 4. Limpiar IndexedDB (offline storage)
                updateStatus('Limpiando bases de datos...');
                if ('indexedDB' in window) {
                    try {
                        await indexedDB.deleteDatabase('offline-viajes-db');
                        log('✓ IndexedDB eliminada');
                    } catch (e) {
                        log('⚠️ No se pudo eliminar IndexedDB: ' + e.message);
                    }
                }

                updateStatus('Esperando 2 segundos...');
                await new Promise(resolve => setTimeout(resolve, 2000));

                log('✓ Limpieza completa');
                showSuccess();

            } catch (error) {
                log('❌ Error: ' + error.message);
                updateStatus('Error en la actualización');
                document.getElementById('spinner').style.display = 'none';
                document.getElementById('homeBtn').style.display = 'inline-block';
            }
        }

        // Ejecutar automáticamente al cargar
        forceUpdate();
    </script>
</body>
</html>
    """

File: static/js/pwa.js
/**
 * PWA Registration - Mi Agente Viajes
 * 
 * Funcionalidades:
 * - Registro del Service Worker
 * - Detección de actualizaciones
 * - Prompt de instalación
 * - Detección online/offline
 */

(function() {
    'use strict';

    // ============================================
    // SERVICE WORKER REGISTRATION
    // ============================================
    
    if ('serviceWorker' in navigator) {
        window.addEventListener('load', async () => {
            try {
                const registration = await navigator.serviceWorker.register('/sw.js', {
                    scope: '/'
                });
                
                console.log('[PWA] Service Worker registered:', registration.scope);
                
                // Verificar actualizaciones
                registration.addEventListener('updatefound', () => {
                    const newWorker = registration.installing;
                    console.log('[PWA] New Service Worker installing...');
                    
                    newWorker.addEventListener('statechange', () => {
                        if (newWorker.state === 'installed' && navigator.serviceWorker.controller) {
                            showUpdateNotification();
                        }
                    });
                });
                
                // Verificar actualizaciones periódicamente (cada hora)
                setInterval(() => {
                    registration.update();
                }, 60 * 60 * 1000);
                
            } catch (error) {
                console.error('[PWA] Service Worker registration failed:', error);
            }
        });
        
        // Recargar cuando el nuevo SW tome control
        let refreshing = false;
        navigator.serviceWorker.addEventListener('controllerchange', () => {
            if (!refreshing) {
                refreshing = true;
                window.location.reload();
            }
        });
    }

    // ============================================
    // UPDATE NOTIFICATION
    // ============================================
    
    function showUpdateNotification() {
        const notification = document.createElement('div');
        notification.id = 'pwa-update-notification';
        notification.innerHTML = `
            <style>
                #pwa-update-notification {
                    position: fixed;
                    bottom: 24px;
                    left: 50%;
                    transform: translateX(-50%);
                    background: #1D1D1F;
                    color: white;
                    padding: 16px 24px;
                    border-radius: 12px;
                    box-shadow: 0 4px 24px rgba(0,0,0,0.2);
                    display: flex;
                    align-items: center;
                    gap: 16px;
                    z-index: 10000;
                    font-family: -apple-system, BlinkMacSystemFont, 'SF Pro Display', 'Segoe UI', Roboto, sans-serif;
                    font-size: 1rem;
                    animation: slideUp 0.3s ease;
                }
                @keyframes slideUp {
                    from { transform: translateX(-50%) translateY(100px); opacity: 0; }
                    to { transform: translateX(-50%) translateY(0); opacity: 1; }
                }
                #pwa-update-notification button {
                    background: #0071E3;
                    color: white;
                    border: none;
                    padding: 8px 16px;
                    border-radius: 8px;
                    font-size: 0.9rem;
                    font-weight: 500;
                    cursor: pointer;
                    transition: background 0.2s;
                }
                #pwa-update-notification button:hover {
                    background: #0077ED;
                }
                #pwa-update-notification .dismiss {
                    background: transparent;
                    padding: 8px;
                    opacity: 0.6;
                }
                #pwa-update-notification .dismiss:hover {
                    opacity: 1;
                    background: rgba(255,255,255,0.1);
                }
            </style>
            <span>Nueva version disponible</span>
            <button onclick="updateApp()">Actualizar</button>
            <button class="dismiss" onclick="this.parentElement.remove()">x</button>
        `;
        
        document.body.appendChild(notification);
    }
    
    window.updateApp = function() {
        if (navigator.serviceWorker.controller) {
            navigator.serviceWorker.controller.postMessage({ type: 'SKIP_WAITING' });
        }
        window.location.reload();
    };

    // ============================================
    // INSTALL PROMPT (A2HS)
    // ============================================
    
    let deferredPrompt = null;
    
    window.addEventListener('beforeinstallprompt', (e) => {
        console.log('[PWA] Install prompt available');
        e.preventDefault();
        deferredPrompt = e;
        
        const installBtn = document.getElementById('pwa-install-btn');
        if (installBtn) {
            installBtn.style.display = 'inline-flex';
        }
    });
    
    window.addEventListener('appinstalled', () => {
        console.log('[PWA] App installed');
        deferredPrompt = null;
        
        const installBtn = document.getElementById('pwa-install-btn');
        if (installBtn) {
            installBtn.style.display = 'none';
        }
    });
    
    window.installPWA = async function() {
        if (!deferredPrompt) {
            console.log('[PWA] No install prompt available');
            return false;
        }
        
        deferredPrompt.prompt();
        const { outcome } = await deferredPrompt.userChoice;
        console.log('[PWA] Install prompt outcome:', outcome);
        deferredPrompt = null;
        
        return outcome === 'accepted';
    };
    
    window.isPWAInstalled = function() {
        return window.matchMedia('(display-mode: standalone)').matches ||
               window.navigator.standalone === true;
    };

    // ============================================
    // ONLINE/OFFLINE DETECTION
    // ============================================
    
    function updateOnlineStatus() {
        const isOnline = navigator.onLine;
        document.body.classList.toggle('offline', !isOnline);
        
        window.dispatchEvent(new CustomEvent('connectionchange', { 
            detail: { online: isOnline } 
        }));
        
        if (isOnline) {
            console.log('[PWA] Back online');
            if ('serviceWorker' in navigator && window.SyncManager && 'sync' in window.SyncManager) {
                navigator.serviceWorker.ready.then(registration => {
                    registration.sync.register('sync-viajes');
                });
            }
        } else {
            console.log('[PWA] Gone offline');
        }
    }
    
    window.addEventListener('online', updateOnlineStatus);
    window.addEventListener('offline', updateOnlineStatus);
    updateOnlineStatus();

    console.log('[PWA] Module loaded');
})();
